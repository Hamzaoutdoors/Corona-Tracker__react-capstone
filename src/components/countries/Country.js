/* eslint-disable react/prop-types */
import React from 'react';
import { motion } from 'framer-motion';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowCircleRight } from '@fortawesome/free-solid-svg-icons';

const Country = ({
  id, name, todayNewConfirmed, carteName,
}) => {
  const numberFormat = (x) => x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

  return (
    <motion.div
      className={`div-${id} d-flex flex-column`}
      initial={{
        opacity: 0,
        translateX: id % 2 === 0 ? -50 : 50,
        translateY: -50,
      }}
      animate={{
        opacity: 1,
        translateX: 0,
        translateY: 0,
      }}
      transition={{
        duration: 0.3,
        delay: id * 0.3,
      }}
    >
      <FontAwesomeIcon icon={faArrowCircleRight} className="arrow-icon" />
      <div className="country-image">
        <img alt="country-img" src={`./cartes/${carteName}.svg`} />
      </div>
      <div
        className="country-body"
      >
        <h2
          className="p-0"
          data-testid="country-test"
        >
          {name}

        </h2>
        <p className="p-0" data-content={`${numberFormat(todayNewConfirmed)} Cases`}>{`${numberFormat(todayNewConfirmed)} Cases`}</p>
      </div>
    </motion.div>
  );
};

export default Country;
