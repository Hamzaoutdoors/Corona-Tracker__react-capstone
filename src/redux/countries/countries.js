import fetchCountries from '../../utils/countries-api';

const FETCH_COUNTRIES = 'narrativa/countries/FETCH_COUNTRIES';
const GET_COUNTRIES = 'narrativa/countries/GET_COUNTRIES';
const API_SUCCESS = 'narrativa/countries/API_SUCCESS';
const API_FAILURE = 'narrativa/countries/API_FAILURE';

const initialState = { countries: [], loaded: false };

export const fetchCountriesAction = () => (dispatch) => {
  dispatch({ type: FETCH_COUNTRIES, loaded: false });
  return fetchCountries().then(
    (countries) => {
      dispatch({ type: API_SUCCESS });
      dispatch({ type: GET_COUNTRIES, countries });
    },
    (error) => {
      dispatch({ type: API_FAILURE, error });
    },
  );
};

const countiresReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_COUNTRIES:
      return { countries: [...action.countries], loaded: true };
    default:
      return state;
  }
};

export default countiresReducer;
