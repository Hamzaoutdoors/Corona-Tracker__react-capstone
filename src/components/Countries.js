import React from 'react';
import { shallowEqual, useSelector } from 'react-redux';
import Country from './Country';
import Header from './Header';

const Countries = () => {
  const countries = useSelector((state) => state.countries, shallowEqual);

  const total = countries.reduce((acc, curr) => acc + curr.todayNewConfirmed, 0);

  const countriesMoreAffected = countries.sort((a, b) => b.todayNewConfirmed - a.todayNewConfirmed);
  const countriesList = countriesMoreAffected.map((country) => {
    const {
      id, name, regions, todayNewConfirmed, todayConfirmed,
    } = country;

    return (
      <Country
        key={id}
        id={countries.indexOf(country)}
        carteName={id}
        name={name}
        regions={regions}
        todayConfirmed={todayConfirmed}
        todayNewConfirmed={todayNewConfirmed}
      />
    );
  });

  return (
    <>
      {' '}
      <Header total={total} title="EUROPE" carteName="europe" />
      <h3 className="middle-title">Today New Confirmed by Country</h3>
      <div className="countries-container">
        {countriesList}
      </div>
    </>
  );
};

export default Countries;
