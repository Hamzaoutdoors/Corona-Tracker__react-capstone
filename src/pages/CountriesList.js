import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleDoubleLeft } from '@fortawesome/free-solid-svg-icons';
import Countries from '../components/Countries';

const CountriesList = () => (
  <div>
    <div className="p-2 text-bold">
      {' '}
      <FontAwesomeIcon icon={faAngleDoubleLeft} className="mx-1" />
      2021
    </div>
    {/* Counties list */}
    <Countries />
  </div>
);

export default CountriesList;
