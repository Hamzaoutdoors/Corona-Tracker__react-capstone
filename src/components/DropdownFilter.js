import { ButtonGroup, Dropdown, DropdownButton } from 'react-bootstrap';
import { useDispatch } from 'react-redux';
import { fetchCountryAction } from '../redux/countries/countries';

const DropdownFilter = () => {
  const dispatch = useDispatch();

  const europecountries = ['Spain', 'Germany', 'Italy', 'France', 'Denmark', 'Portugal', 'Netherlands', 'Switzerland', 'United_Kingdom'];
  const buttonList = europecountries.map((country) => <Dropdown.Item className="text-center" key={country} eventKey={country} data-testid={country}>{country}</Dropdown.Item>);

  const filterHandler = (e) => {
    dispatch(fetchCountryAction(e.toLowerCase()));
  };

  return (

    <DropdownButton className="w-50 m-2" variant="warning" id="dropdown-button-drop-end" drop="end" title="Filter By Country" key="end" as={ButtonGroup} onSelect={filterHandler}>
      {buttonList}
    </DropdownButton>

  );
};

export default DropdownFilter;
