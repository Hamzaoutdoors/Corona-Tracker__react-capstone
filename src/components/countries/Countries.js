import React from 'react';
import { Link } from 'react-router-dom';
import { shallowEqual, useSelector } from 'react-redux';
import Country from './Country';
import Header from '../Header';
import Loading from '../Loading';
import DropdownFilter from '../DropdownFilter';

// import Loading from './Loading';

const Countries = () => {
  const { countries } = useSelector((state) => state.countries, shallowEqual);
  const total = countries.reduce((acc, curr) => acc + curr.todayNewConfirmed, 0);
  const countriesMoreAffected = countries.sort((a, b) => b.todayNewConfirmed - a.todayNewConfirmed);

  const countriesList = countriesMoreAffected.map((country) => {
    const {
      id, name, regions, todayNewConfirmed, todayConfirmed,
    } = country;

    return (
      <Link
        key={id}
        to={`/country/${id}`}
        className="div"
        data-testid={`button-${id}`}
      >
        <Country
          id={countries.indexOf(country)}
          carteName={id}
          name={name}
          regions={regions}
          todayConfirmed={todayConfirmed}
          todayNewConfirmed={todayNewConfirmed}
        />
      </Link>
    );
  });

  return (
    <div>
      {' '}
      <Header total={`${total} Cases`} title="EUROPE" carteName="europe" />
      <h3 className="middle-title">Today New Confirmed by Country</h3>
      {countries.length === 0 ? <Loading /> : <DropdownFilter />}
      <div className="countries-container">
        {countriesList}
      </div>
    </div>
  );
};

export default Countries;
