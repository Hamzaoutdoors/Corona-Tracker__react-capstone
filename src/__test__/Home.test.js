import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  render, fireEvent, screen,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/configureStore';
import Countries from '../components/countries/Countries';
import CountriesList from '../pages/CountriesList';
import DropdownFilter from '../components/DropdownFilter';
import { fetchCountriesAction } from '../redux/countries/countries';
import '@testing-library/jest-dom';

jest.mock('../utils/countries-api');

describe('Home page tests', () => {
  test('Countries page matches snapshot', async () => {
    await store.dispatch(fetchCountriesAction());
    const countriesPage = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Countries />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(countriesPage).toMatchSnapshot();
  });

  describe('display europe countries', () => {
    test('Countries page interaction', () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <Countries />
          </BrowserRouter>
        </Provider>,
      );
      screen.queryAllByRole('link').forEach((role) => expect(role).toBeInTheDocument());
    });
    test('Fire filter event', () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <CountriesList />
          </BrowserRouter>
        </Provider>,
      );
      fireEvent.select(screen.getByTestId('button-germany'));
      expect(screen.getByText('Germany')).toBeInTheDocument();
    });
  });
  test('dropdown displayed properly ', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <DropdownFilter />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
