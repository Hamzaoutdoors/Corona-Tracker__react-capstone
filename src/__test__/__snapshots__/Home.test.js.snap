// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page tests Countries page matches snapshot 1`] = `
<div>
   
  <div
    className="page-header"
  >
    <div
      className="header-image"
    >
      <img
        alt="cart"
        src="/cartes/europe.svg"
      />
    </div>
    <div
      className="header-body"
    >
      <h2>
        EUROPE
      </h2>
      <p>
        5030 Cases
      </p>
    </div>
  </div>
  <h3
    className="middle-title"
  >
    Today New Confirmed by Country
  </h3>
  <div
    className="w-50 m-2 dropend btn-group"
    role="group"
  >
    <button
      aria-expanded={false}
      className="dropdown-toggle btn btn-warning"
      disabled={false}
      id="dropdown-button-drop-end"
      onClick={[Function]}
      type="button"
    >
      Filter By Country
    </button>
  </div>
  <div
    className="countries-container container"
  >
    <a
      className="div"
      data-testid="button-france"
      href="/country/france"
      onClick={[Function]}
    >
      <div
        className="div-0 d-flex flex-column"
        style={
          Object {
            "opacity": 0,
            "transform": "translateX(-50px) translateY(-50px) translateZ(0)",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-arrow-circle-right fa-w-16 arrow-icon"
          data-icon="arrow-circle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <div
          className="country-image"
        >
          <img
            alt="country-img"
            src="./cartes/france.svg"
          />
        </div>
        <div
          className="country-body"
        >
          <h2
            className="p-0"
            data-testid="country-test"
          >
            France
          </h2>
          <p
            className="p-0"
            data-content="3,991 Cases"
          >
            3,991 Cases
          </p>
        </div>
      </div>
    </a>
    <a
      className="div"
      data-testid="button-italy"
      href="/country/italy"
      onClick={[Function]}
    >
      <div
        className="div-1 d-flex flex-column"
        style={
          Object {
            "opacity": 0,
            "transform": "translateX(50px) translateY(-50px) translateZ(0)",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-arrow-circle-right fa-w-16 arrow-icon"
          data-icon="arrow-circle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <div
          className="country-image"
        >
          <img
            alt="country-img"
            src="./cartes/italy.svg"
          />
        </div>
        <div
          className="country-body"
        >
          <h2
            className="p-0"
            data-testid="country-test"
          >
            Italy
          </h2>
          <p
            className="p-0"
            data-content="1,000 Cases"
          >
            1,000 Cases
          </p>
        </div>
      </div>
    </a>
    <a
      className="div"
      data-testid="button-germany"
      href="/country/germany"
      onClick={[Function]}
    >
      <div
        className="div-2 d-flex flex-column"
        style={
          Object {
            "opacity": 0,
            "transform": "translateX(-50px) translateY(-50px) translateZ(0)",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-arrow-circle-right fa-w-16 arrow-icon"
          data-icon="arrow-circle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <div
          className="country-image"
        >
          <img
            alt="country-img"
            src="./cartes/germany.svg"
          />
        </div>
        <div
          className="country-body"
        >
          <h2
            className="p-0"
            data-testid="country-test"
          >
            Germany
          </h2>
          <p
            className="p-0"
            data-content="39 Cases"
          >
            39 Cases
          </p>
        </div>
      </div>
    </a>
  </div>
</div>
`;

exports[`Home page tests dropdown displayed properly  1`] = `
<div
  className="w-50 m-2 dropend btn-group"
  role="group"
>
  <button
    aria-expanded={false}
    className="dropdown-toggle btn btn-warning"
    disabled={false}
    id="dropdown-button-drop-end"
    onClick={[Function]}
    type="button"
  >
    Filter By Country
  </button>
</div>
`;
